diff --git a/Makefile b/Makefile
index da92533..46d7e14 100644
--- a/Makefile
+++ b/Makefile
@@ -354,6 +354,17 @@ else ifeq ($(platform), miyoo)
 	CFLAGS += -funsafe-math-optimizations -fsingle-precision-constant -fexpensive-optimizations
 	CFLAGS += -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-unroll-loops
 
+# SF2000
+else ifeq ($(platform), sf2000)
+	TARGET := $(TARGET_NAME)_libretro_$(platform).a
+	MIPS:=/opt/mips32-mti-elf/2019.09-03-2/bin/mips-mti-elf-
+	CC = $(MIPS)gcc
+	AR = $(MIPS)ar
+	CFLAGS = -EL -march=mips32 -mtune=mips32 -msoft-float -G0 -mno-abicalls -fno-pic
+	CFLAGS += -ffast-math -fomit-frame-pointer -ffunction-sections -fdata-sections
+	CFLAGS += -DSF2000 -DLOWRES -DM16B
+	STATIC_LINKING = 1
+
 # emscripten
 else ifeq ($(platform), emscripten)
 	TARGET := $(TARGET_NAME)_libretro_$(platform).bc
diff --git a/cap32/crtc.c b/cap32/crtc.c
index 096bccb..588f9a2 100644
--- a/cap32/crtc.c
+++ b/cap32/crtc.c
@@ -1241,6 +1241,7 @@ void crtc_cycle(int repeat_count)
             RendPos = RendStart;
             HorzChar--;
          } else {
+            val = val & ~3;		// osaka's misaligned access fix
             RendPos = (uint32_t *)&RendBuff[val];
             int tmp = (uint8_t*)RendStart - (uint8_t*)RendPos;
             HorzPix[48] = (uint8_t)tmp;
