diff --git a/.gitignore b/.gitignore
index ddaa784..8bf71e2 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,3 +4,5 @@ android-libs/launcher/src/com/beloko/idtech/R.java
 /*build*/
 /results/
 /.idea/
+*.o
+*.a
diff --git a/src/id_ca.cpp b/src/id_ca.cpp
index 358580a..910c270 100644
--- a/src/id_ca.cpp
+++ b/src/id_ca.cpp
@@ -75,14 +75,14 @@ void CA_CacheMap (const FString &mapname, bool loading)
 	static TUniquePtr<GameMap, TFuncDeleter<GameMap, CA_UnloadMap> > map;
 	map.Reset();
 
-	Printf("\n");
+	//Printf("\n");
 
 	strncpy(gamestate.mapname, mapname, 8);
 	levelInfo = &LevelInfo::Find(mapname);
 	::map = map = new GameMap(mapname);
 	map->LoadMap(loading);
 
-	Printf("\n%s - %s\n\n", mapname.GetChars(), levelInfo->GetName(map).GetChars());
+	Printf("CA_CacheMap: map=%s level='%s'\n", mapname.GetChars(), levelInfo->GetName(map).GetChars());
 
 	CalcVisibility(gLevelVisibility);
 }
diff --git a/src/libretro/Makefile b/src/libretro/Makefile
index ab06c30..753b886 100644
--- a/src/libretro/Makefile
+++ b/src/libretro/Makefile
@@ -370,6 +370,21 @@ else ifeq ($(platform), miyoo)
 	COMMON_CFLAGS += -fomit-frame-pointer -ffast-math -march=armv5te -mtune=arm926ej-s
 	COMMON_CFLAGS += -fno-common -ftree-vectorize -funswitch-loops
    COMMON_CFLAGS += -Dstricmp=strcasecmp -Dstrnicmp=strncasecmp
+
+# SF2000
+else ifeq ($(platform), sf2000)
+	TARGET := $(TARGET_NAME)_libretro_$(platform).a
+	MIPS=/opt/mips32-mti-elf/2019.09-03-2/bin/mips-mti-elf-
+	CC = $(MIPS)gcc
+	CXX = $(MIPS)g++
+	AR = $(MIPS)ar
+	CFLAGS = -EL -march=mips32 -mtune=mips32 -msoft-float -G0 -mno-abicalls -fno-pic
+	CFLAGS += -ffast-math -fomit-frame-pointer -ffunction-sections -fdata-sections
+	CFLAGS += -DSF2000 -DHAVE_STRL -DUSE_OWN_ADDED_SIZE
+	CXXFLAGS := $(CFLAGS) -fno-use-cxa-atexit
+	COMMON_CFLAGS += -Dstricmp=strcasecmp -Dstrnicmp=strncasecmp
+	STATIC_LINKING = 1
+
 else
    CC ?= gcc
    TARGET := $(TARGET_NAME)_libretro.dll
@@ -442,7 +457,7 @@ COMMON_CFLAGS   += -Wall -DLIBRETRO -D__LIBRETRO__ $(fpic) -DCORE=ON -DSDL2=ON -
 
 ifeq (,$(findstring msvc,$(platform)))
   COMMON_CFLAGS += -ffast-math -fno-strict-aliasing -fvisibility=hidden -fomit-frame-pointer -Wno-sign-compare -Wno-switch -Wno-format-security
-  CXXFLAGS += -fpermissive -std=gnu++98
+  CXXFLAGS += -fpermissive -std=gnu++11
 ifneq ($(platform), emscripten)
   COMMON_CFLAGS += -fno-unsafe-math-optimizations
 endif
diff --git a/src/libretro/libretro-common b/src/libretro/libretro-common
index 996376e..fce57fd 160000
--- a/src/libretro/libretro-common
+++ b/src/libretro/libretro-common
@@ -1 +1 @@
-Subproject commit 996376e36d3f4f56eba202cb96230568628d2583
+Subproject commit fce57fdfb7514a62540781aecbc4a53aedee3dce
diff --git a/src/libretro/libretro.cpp b/src/libretro/libretro.cpp
index b6ad5cf..823123a 100644
--- a/src/libretro/libretro.cpp
+++ b/src/libretro/libretro.cpp
@@ -398,7 +398,7 @@ void libretro_log(const char *format, ...)
 
 	va_start(va, format);
 	vsnprintf(formatted, sizeof(formatted) - 1, format, va);
-	log_cb(RETRO_LOG_INFO, "%s\n", formatted);
+	log_cb(RETRO_LOG_INFO, "%s", formatted);
 	va_end(va);
 }
 
@@ -406,7 +406,7 @@ void Quit ()
 {
 	struct retro_message msg;
 
-	libretro_log("Fatal error");
+	libretro_log("Fatal error\n");
 	msg.msg    = "Fatal error";
 	msg.frames = fp10s;
 	environ_cb(RETRO_ENVIRONMENT_SET_MESSAGE, &msg);
@@ -596,14 +596,14 @@ static void am_multiple_choice (const char *name, unsigned &var, bool &is_update
 
 static void update_variables(bool startup)
 {
-#if defined(RS90)
+#if defined(RS90) || defined(SF2000)
 	store_files_in_memory = false;
 #elif defined(_3DS) || defined(GEKKO)
 	store_files_in_memory = true;
 #else
 	store_files_in_memory = get_bool_option("ecwolf-memstore");
 #endif
-#ifdef RS90
+#if defined(RS90) || defined(SF2000)
 	preload_digital_sounds = false;
 #else
 	preload_digital_sounds = get_bool_option("ecwolf-preload-digisounds");
@@ -1028,7 +1028,7 @@ static void TransformInputs(wl_input_state_t *input)
 		break;
 	}
 
-	if((newly_pressed & (1<<RETRO_DEVICE_ID_JOYPAD_SELECT))) {
+	if(map && (newly_pressed & (1<<RETRO_DEVICE_ID_JOYPAD_SELECT))) {
 		AM_Toggle();
 		return;
 	}
@@ -1113,7 +1113,7 @@ static void mixChannel(long long tic, SoundChannelState *channel)
 #define MB(x) ((x) << 20)
 
 size_t limit_sound_cache_size =
-#ifdef RS90
+#if defined(RS90) || defined(SF2000)
 	MB(5)
 #else
 	MB(15)
@@ -1134,7 +1134,7 @@ void generate_audio(long long tic)
 		// We don't want to keep dropping and reloading the same files every frame
 		if (limit_sound_cache_size <= (touched_sound_size * 3) / 2) {
 			limit_sound_cache_size = (touched_sound_size * 3) / 2;
-#ifdef RS90
+#if defined(RS90) || defined(SF2000)
 			if (limit_sound_cache_size >= MB(7))
 				limit_sound_cache_size = MB(7);
 #endif		
@@ -1156,6 +1156,8 @@ void generate_silent_audio(void)
 
 void retro_run(void)
 {
+frame_time_cb(0);
+
 	int expectframes = 0;
 	wl_input_state_t input;
 	long long framestarttic = GetTimeCount();
diff --git a/src/libretro/libretro_core_options.h b/src/libretro/libretro_core_options.h
index 8697c0d..5489c6b 100644
--- a/src/libretro/libretro_core_options.h
+++ b/src/libretro/libretro_core_options.h
@@ -108,6 +108,8 @@ struct retro_core_option_definition option_defs_us[] = {
 		},
 #if defined(RS90)
 		"240x160",
+#elif defined(SF2000)
+		"320x240",
 #elif defined(VITA) || defined(PSP)
 		"480x270",
 #elif defined (_3DS)
diff --git a/src/lumpremap.cpp b/src/lumpremap.cpp
index 86365bf..cf23a59 100644
--- a/src/lumpremap.cpp
+++ b/src/lumpremap.cpp
@@ -186,7 +186,7 @@ bool LumpRemapper::LoadMap()
 	int lump = Wads.GetNumForName(mapLumpName);
 	if(lump == -1)
 	{
-		printf("\n");
+		printf("LumpRemapper::LoadMap - Wads.GetNumForName failed. mapLumpName=%s\n", mapLumpName.GetChars());
 		return false;
 	}
 	FWadLump mapLump = Wads.OpenLumpNum(lump);
diff --git a/src/wl_def.h b/src/wl_def.h
index 0d05700..6f4657f 100644
--- a/src/wl_def.h
+++ b/src/wl_def.h
@@ -20,6 +20,9 @@
 #include <minwindef.h>
 #endif
 
+extern "C" void xlog(const char *format, ...);
+#define printf(...) xlog(__VA_ARGS__)
+
 #ifdef _MSC_VER
 #define PACKED
 #define PACK_START __pragma(pack(push, 1))
diff --git a/src/zdoomsupport.h b/src/zdoomsupport.h
index 4b655bb..a8a3c9b 100644
--- a/src/zdoomsupport.h
+++ b/src/zdoomsupport.h
@@ -9,8 +9,10 @@
 #include <cstring>
 
 #if defined(LIBRETRO) && LIBRETRO
+extern "C" void xlog(const char *format, ...);
 void libretro_log(const char *format, ...);
-#define printf(...) libretro_log(__VA_ARGS__)
+// #define printf(...) libretro_log(__VA_ARGS__)
+#define printf(...) xlog(__VA_ARGS__)
 #elif defined(__ANDROID__)
 #include <android/log.h>
 #define printf(...) __android_log_print(ANDROID_LOG_INFO,"ECWolf",__VA_ARGS__)
